src/test/datascience/debugger.vscode.test.ts

```diff
@@ -0,0 +1,160 @@
+        sinon.restore();
+        commandManager = api.serviceContainer.get<ICommandManager>(ICommandManager);
+        const coreVariableViewProvider = api.serviceContainer.get<IVariableViewProvider>(IVariableViewProvider);
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        variableViewProvider = (coreVariableViewProvider as any) as ITestVariableViewProvider; // Cast to expose the test interfaces
+        debuggingManager = api.serviceContainer.get<IDebuggingManager>(IDebuggingManager);
+        vscodeNotebook = api.serviceContainer.get<IVSCodeNotebook>(IVSCodeNotebook);
+        traceInfo(`Start Test Suite (completed)`);
+    });
+    setup(async function () {
+        traceInfo(`Start Test ${this.currentTest?.title}`);
+        sinon.restore();
+
+        // Create an editor to use for our tests
+        await createEmptyPythonNotebook(disposables);
+        traceInfo(`Start Test (completed) ${this.currentTest?.title}`);
+    });
+    teardown(async function () {
+        traceInfo(`Ended Test ${this.currentTest?.title}`);
+        await closeNotebooks(disposables);
+        await closeNotebooksAndCleanUpAfterTests(disposables);
+        traceInfo(`Ended Test (completed) ${this.currentTest?.title}`);
+    });
+
+    // Cleanup after suite is finished
+    suiteTeardown(() => closeNotebooksAndCleanUpAfterTests(disposables));
+
+    test('Run by Line - Full Workflow', async function () {
+        // set up
+        await insertCodeCell('a=1\na', { index: 0 });
+        const doc = vscodeNotebook.activeNotebookEditor?.document!;
+        const cell = doc.getCells()![0]!;
+
+        // Start Run by Line
+        await commandManager.executeCommand(Commands.RunByLine, cell);
+
+        // Check that a debugging session is created
+        const session = debuggingManager.getDebugSession(doc);
+        assert.isOk<DebugSession | undefined>(session, 'Session not started');
+
+        // Check that the debug adapter is created
+        const debugAdapter = debuggingManager.getDebugAdapter(doc);
+        assert.isOk<KernelDebugAdapter | undefined>(debugAdapter, 'DebugAdapter not started');
+
+        // Wait for the stoped event
+        let msg = await waitForEvent<DebugProtocol.StoppedEvent>('stopped', debugAdapter!);
+
+        // Check that we're stopped on the cell
+        const stack = await session!.customRequest('stackTrace', {
+            threadId: msg.body.threadId
+        });
+        assert.isTrue(stack.stackFrames.length > 0, 'has frames');
+        assert.equal(stack.stackFrames[0].source?.path!, cell.document.uri.toString(), 'Stopped at the worng path');
+
+        // continue to the next line
+        await commandManager.executeCommand(Commands.RunByLineContinue, cell);
+
+        // Wait for the stoped event
+        msg = await waitForEvent<DebugProtocol.StoppedEvent>('stopped', debugAdapter!);
+        // Wait for variable view to be updated
+        await sleep(1000);
```

Comment:
Is there an event on the variable view that you can listen to?

Reply:
it was tricky, but yeah there is